### PR TITLE
feat: add fund rebalancing cost cap

### PR DESCRIPTION
**Pull request by nadim-az**

### CODE REVIEW

File: config/local/config.yml
Comment (mattac21) [inline config/local/config.yml:57]: i feel like these should be moved to the fund rebalancer config and then drop the `fund_rebalancer_` prefix

File: fundrebalancer/fundrebalancer.go
Comment (mattac21) [inline fundrebalancer/fundrebalancer.go:692]: you are checking if `chainConfig.FundRebalancingTimeout` is nil here, but above you have already dereferenced it, which implies that the dereference on line 683 could be a nil pointer dereference

File: fundrebalancer/fundrebalancer.go
Comment (dhfang) [inline fundrebalancer/fundrebalancer.go:705]: I dont believe this has the desired effect. This logic uses the creation time of the oldest pending transfer, but a pending transfer has already landed on chain, so I'm not sure this gives the insight we want into whether the solver rebalance has been blocked on meeting the max cost check. We should do something like tracking in memory how long the rebalancer has been failing on profitability checks per chain and using that to determine the timeout.

File: shared/config/config.go
Comment (mattac21) [inline shared/config/config.go:110]: if setting to -1 will disable the timeout, what is the reason to have this as a pointer? a 0ns timeout doesn't really make sense so we could probably just enforce in the config validation that the timeout isn't 0 (which will be the 0 value if it isn't a pointer), instead of having it as a pointer and checking that it is not nil

Comment (mattac21) [reply shared/config/config.go:110]: this would also fix my above comment about the dereference before the nil check

File: fundrebalancer/fundrebalancer.go
Comment (mattac21) [inline fundrebalancer/fundrebalancer.go:702]: i may just be misunderstanding here so lmk, but I don't entirely get why we are using the oldest pending transfers timestamp to determine if the rebalance we are trying to do has been tried before and it timed out due to gas being too high. 

the chain id in this function is the chain that they are rebalancing from (i.e funds are being sent from this chainID to another chain), unless I am misunderstanding, this chain id won't have rebalances _to_ it (chainID will always be an evm chain which we don't rebalance to) so the `r.database.GetPendingRebalanceTransfersToChain` would always be an empty list i think

also assuming im wrong/misunderstanding about the above and we do get back a list of transfers from this chain, then what if there just are no pending transfers? a pending transfer will only be in the db if a transfer has actually been initiated on chain (and therefore passed the gas check). it seems like if there are no pending transfers in the db (because gas is too high) the transfer the solver is trying to do will never time out because there are no transfers in the db to check the ts against.

maybe we should be inserting txs to the db in some `pending_gas_approval` status, so we can know when we started trying to send the transfer and therefore know when it has timed out due to gas not being low enough (and then apply the cost cap)?

**Review by mattac21 — APPROVED** [review]
this looks good to me! I would let dennis review as well tho before merging. I think the only other thing I would maybe add is a comment in config.go for the timeout saying that this is tracked in memory for a transfer, so if the solver restarts this timeout is set. I think if I was running a solver I would assume that solver restarts don't reset the timeout clock, so this may be unexpected.

**Review by dhfang — APPROVED** [review]
LGTM! I think a big part of why this is confusing is just config naming. We have a two tier profit check design, which if presented as a tier 1 price, a tier 1 timeout, and a tier 2 price is not that confusing imo albeit complicated still. We currently have tier 1 named "max" and still have a tier 2 that's higher than that, which is confusing. We should definitely come back and clean this up, which potentially involves a rethinking of the entire profit checking design, but we can handhold solvers for now imo.

**Pull request merged**
